pkg/util/concurrency/runner.go

```diff
@@ -0,0 +1,64 @@
+package concurrency
+
+import (
+	"context"
+	"sync"
+
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+)
+
+// ForeachUser runs the provided userFunc for each userIDs up to concurrency concurrent workers.
+// In case userFunc returns error, it will continue to process remaining users but returns an
+// error with all errors userFunc has returned.
+func ForeachUser(ctx context.Context, userIDs []string, concurrency int, userFunc func(userID string) error) error {
+	wg := sync.WaitGroup{}
+	ch := make(chan string)
+
+	// Keep track of all errors occurred.
+	errs := tsdb_errors.NewMulti()
+	errsMx := sync.Mutex{}
+
+	for ix := 0; ix < concurrency; ix++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for userID := range ch {
+				// Ensure the context has not been canceled (ie. shutdown has been triggered).
+				if ctx.Err() != nil {
+					break
+				}
+
+				if err := userFunc(userID); err != nil {
```

Comment:
Should we pass context to the `userFunc` instead of passing it via local variables (as callers of `ForEachUser` do now)?

Reply:
Yes, I think it's a better approach.